compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -3324,19 +3332,22 @@ fn print_disambiguation_help<'tcx>(
                 .get(0)
                 .and_then(|ty| ty.ref_mutability())
                 .map_or("", |mutbl| mutbl.ref_prefix_str());
-            let args = format!(
-                "({}{})",
-                rcvr_ref,
-                std::iter::once(receiver)
-                    .chain(args.iter())
-                    .map(|arg| tcx
-                        .sess
-                        .source_map()
-                        .span_to_snippet(arg.span)
-                        .unwrap_or_else(|_| { "_".to_owned() }))
-                    .collect::<Vec<_>>()
-                    .join(", "),
-            );
+            // If the type of first arg of this assoc function is `Self` or current trait impl type, we need to take the receiver as args. Otherwise, we don't.
+            let args = if let Some(t) = first_arg_type
+                && (t.to_string() == "Self" || t == trait_impl_type)
```

Comment:
Without having checked if the rest of your implementation makes sense, I'll suggest the following:

```suggestion
                && (t == tcx.types.self_param || t == trait_impl_type)
```